### ISSUE TITLE
false is indistinguishable from NULL in a tuple of bools

### ISSUE THREAD

**Issue by sumerman** [tcdi/pgx#792]
Consider the snippet below: I would expect `test_foo` to pass, and yet the assertion fails on `0.5.x`.

```rust
    fn from_helper<T: FromDatum + IntoDatum>(d: Datum) -> Option<T> {
        unsafe { T::from_polymorphic_datum(d, false, pg_sys::InvalidOid) }
    }

    #[pg_test]
    fn test_foo() {
        let d = (Some(false), Some(true)).into_datum().unwrap();
        let (a, b) = from_helper::<(Option<bool>, Option<bool>)>(d).unwrap();
        assert_eq!(a.is_some(), true);
    }
```

I discovered this behavior while trying to make `Spi::get_three_with_args` work in our tests. `Spi::get_one_with_args` works just fine, but `get_two` and `get_three` both suffer from this quirk. I managed to narrow it down to `Spi` functions round-tripping datums in order to move allocation into an appropriate memory context and thus relying on a bunch of `IntoDatum` and `FromDatum` instances starting from `(Option<A>, Option<B>)`.

**Comment by sumerman** [comment]
It would seem something is up with `Vec<Option<T>>` Into and From instances:

```rust
    #[pg_test]
    fn test_bar() {
        let d = vec![Some(false).into_datum(), Some(true).into_datum()].into_datum().unwrap();
        let a = from_helper::<Vec<Option<pg_sys::Datum>>>(d).unwrap();
        assert_eq!(a.first().unwrap().is_some(), true); // fails; a is [None, Some(Datum(...))]
    }
```

**Comment by eeeebbbbrrrr** [comment]
Good catch.  I think the FromDatum impls in `tuples.rs` are bugged.  Looks like they've been bugged for a few years.  

I'll put up a PR now.

**Comment by sumerman** [comment]
> Good catch. I think the FromDatum impls in `tuples.rs` are bugged. Looks like they've been bugged for a few years.
> 
> I'll put up a PR now.

Thank you! It's odd to hear, as it used to work on 0.4.x branch

**Comment by eeeebbbbrrrr** [comment]
> Thank you! It's odd to hear, as it used to work on 0.4.x branch

Yeah, apparently it's not quite what I thought it was.  I'm digging into it now...

**Comment by eeeebbbbrrrr** [comment]
It gets more fun... this fails too:

```rust
    #[pg_test]
    fn test_vec_ints() {
        let d = vec![Some(0).into_datum(), Some(1).into_datum()].into_datum().unwrap();
        let a = from_helper::<Vec<Option<pg_sys::Datum>>>(d).unwrap();
        assert_eq!(a.first().unwrap().is_some(), true);
    }
```

There's something up with pgx thinking a Datum with a value of zero is actually null.  Gonna have to ping @workingjubilee, I think.

**Comment by eeeebbbbrrrr** [comment]
I think the problem really comes from this:

https://github.com/tcdi/pgx/blob/18955c7b46afa3e1140433206ebba3929c7d008b/pgx-pg-sys/src/submodules/datum.rs#L60-L63

A `pgx:Datum` itself has no concept of NULL.  It just exists.  To know if it's NULL or not, Postgres passes around an extra boolean argument.

As such, when the `pgx::Datum`'s value is zero, it's being interpreted as null.

Removing that function and its usages throughout the code fixes this problem.  I think this is the right answer.

**Comment by sumerman** [comment]
> I think the problem really comes from this:
> 
> https://github.com/tcdi/pgx/blob/18955c7b46afa3e1140433206ebba3929c7d008b/pgx-pg-sys/src/submodules/datum.rs#L60-L63
> 
> A `pgx:Datum` itself has no concept of NULL. It just exists. To know if it's NULL or not, Postgres passes around an extra boolean argument.
> 
> As such, when the `pgx::Datum`'s value is zero, we're interpreting it as NULL.
> 
> Removing that function and its usages throughout the code fixes this problem. I think this is the right answer.

🤔 since starting with 0.5.x `Datum` is a newtype we'd still need a way to check whether it contains a null-pointer when treated like a pointer.

**Comment by eeeebbbbrrrr** [comment]
"when treated like a pointer" is the key.

`FromDatum::from_datum()` has an `is_null` argument, which is what we ought to be using.  I'm still trying to sort this out as the above did light up a test that is now crashing Postgres.

**Comment by eeeebbbbrrrr** [comment]
Well, if we accept the code as it is, then the problem is really just here:  https://github.com/tcdi/pgx/blob/18955c7b46afa3e1140433206ebba3929c7d008b/pgx/src/datum/from.rs#L128-L142

Line 136 should just read `if is_null {`.

That fixes your tests, the few extra I've made, and doesn't light up any others.

**Comment by eeeebbbbrrrr** [comment]
PR here:  https://github.com/tcdi/pgx/pull/795

**Comment by workingjubilee** [comment]
SQL's `NULL` and `nullptr_t` are not the same thing and at no point should the same word be used for them, but unfortunately our documentation and interfaces are rife with this conceptual overload anyways. I am even doubtful about our use of `Option<T>` to represent nullable values, as that overloads with several other classical meanings of `Option<T>`.

**Comment by workingjubilee** [comment]
This has been fixed in 0.5.4 which is out now.
